src/artemis/nexus_writing/write_nexus.py

```diff
@@ -0,0 +1,240 @@
+
+    Args:
+        grid_scan_params (GridScanParams): The grid scan parameters.
+
+    Returns:
+        Spec: A scanspec for nexgen
+    """
+    y_line = Line(
+        "sam_y",
+        grid_scan_params.y1_start,
+        grid_scan_params.y_end,
+        grid_scan_params.y_steps
+        + 1,  # 1 more as we take an image on the first step as well as the last
+    )
+    x_line = Line(
+        "sam_x",
+        grid_scan_params.x_start,
+        grid_scan_params.x_end,
+        grid_scan_params.x_steps + 1,
+    )
+    return y_line * ~x_line
+
+
+class NexusWriter:
+    def __init__(self, parameters: FullParameters) -> None:
+        self.detector = create_detector_parameters(parameters.detector_params)
+        self.goniometer = create_goniometer_axes(parameters.detector_params)
+        self.beam, self.attenuator = create_beam_and_attenuator_parameters(
+            parameters.ispyb_params
+        )
+        self.scan_spec = create_scan_spec(parameters.grid_scan_params)
+        self.directory, self.filename = (
+            Path(parameters.detector_params.directory),
+            parameters.detector_params.prefix,
+        )
+        self.num_of_images = parameters.detector_params.num_images
+        self.nexus_file = self.directory / f"{self.filename}.nxs"
+
+    def _get_current_time(self):
+        return datetime.utcfromtimestamp(time.time()).strftime(r"%Y-%m-%dT%H:%M:%SZ")
+
+    def __enter__(self):
+        """
+        Creates a nexus file based on the parameters supplied when this obect was initialised.
+        """
+        start_time = self._get_current_time()
+
+        scan_range = calculate_scan_from_scanspec(self.scan_spec)
+
+        image_data = [self.directory / f"{self.filename}_000001.h5"]
+        metafile = self.directory / f"{self.filename}_meta.h5"
+
+        with h5py.File(self.nexus_file, "x") as nxsfile:
+            nxentry = write_NXentry(nxsfile)
+
+            nxentry.create_dataset("start_time", data=np.string_(start_time))
+
+            call_writers(
+                nxsfile,
+                image_data,
+                "mcstas",
+                scan_range,
+                ("images", self.num_of_images),
+                self.goniometer,
+                self.detector,
+                module,
+                source,
+                self.beam,
+                self.attenuator,
+                vds="dataset",
+                metafile=metafile,
+                link_list=dset_links,
+            )
+
+            image_vds_writer(
+                nxsfile,
+                (
+                    self.num_of_images,
+                    self.detector["image_size"][1],
+                    self.detector["image_size"][0],
+                ),
+            )
+
+    def __exit__(self, *_):
+        with h5py.File(self.nexus_file, "r+") as nxsfile:
+            nxsfile["entry"].create_dataset(
+                "end_time", data=np.string_(self._get_current_time())
+            )
```

Comment:
If we do go with an external timestamp for start/end time as per my earlier comment I'm not sure the context manager will be necessary, as its main function at the moment seems to be to make sure the end time is written.

Reply:
We'll still need to actually write the `end_time` into the nexus file, even if the actual timestamp comes from elsewhere